lib/sprockets/less/v2/less_template.rb

```diff
@@ -0,0 +1,252 @@
+            include options[:functions] if options[:functions]
+            class_eval(&block) if block_given?
+          end
+        end
+
+        @less_functions_initialized = false
+        class << self
+          attr_accessor :less_functions_initialized
+          alias less_functions_initialized? less_functions_initialized
+          # Templates are initialized once the functions are added.
+          def engine_initialized?
+            less_functions_initialized?
+          end
+        end
+
+        # Add the Less functions if they haven't already been added.
+        def initialize_engine
+          return if self.class.engine_initialized?
+
+          if Sprockets::Less.add_less_functions
+            begin
+              require 'sprockets/helpers'
+              require 'sprockets/less/functions'
+              self.class.less_functions_initialized = true
+            rescue LoadError; end
+          end
+        end
+
+        def call(input)
+          @input = input
+          @filename = input[:filename]
+          @source   = input[:data]
+          @context  = input[:environment].context_class.new(input)
+          run
+        end
+
+        def render(context, _empty_hash_wtf)
+          @context = context
+          run
+        end
+
+
+        def less_engine(data, less_options, css_options)
+          ::Less.Parser['sprockets_context'] = context
+          parser = ::Less::Parser.new(less_options)
+          parser.tree.extend_js @functions
+          engine = parser.parse(data)
+          engine.to_css(css_options)
+        end
+
+        def run
+          css = retrieve_from_cache_store
+          if css.nil?
+            data = Sprockets::Less::Utils.read_file_binary(filename, options)
+            new_data, dependencies = process_dependencies(data)
+
+            css  = less_engine(new_data, less_options, css_options)
+            store_into_cache_store(css)
+
+            less_dependencies = Set.new([filename])
+            if context.respond_to?(:metadata)
+              dependencies.map do |dependency|
+                less_dependencies << dependency.to_s
+                context.metadata[:dependencies] << Sprockets::URIUtils.build_file_digest_uri(dependency.to_s)
+              end
+              context.metadata.merge(data: css, less_dependencies: less_dependencies)
+            else
+              css
+            end
+          else
+            css
+          end
+
+        rescue => e
+          # Annotates exception message with parse line number
+          raise [e, e.backtrace].join("\n")
+        end
+
+        def css_options
+          css_keys = [:compress, :optimization, :silent, :color]
+          Hash[less_options.to_enum.to_a.select{|k, _| css_keys.include? k}]
+        end
+
+
+        def process_dependencies(data)
+          fetch_importer_class.fetch_all_dependencies(data, filename, less_options, css_options)
+        end
+
+        def merge_less_options(options, other_options)
```

Comment:
So this is side-effectful on both `options` and `other_options`? I'd have expected only `options` to suffer from side effects (like Hash#merge!). Maybe some `dup` is in order, or explain the rationale and/or behaviour in a comment.